### ISSUE TITLE
Inconsistent ROL UpdateType flags between processes

### ISSUE THREAD

**Issue by rubyturner** [g-adopt/g-adopt#47]
Flagging an issue a few of my simulations are running into with ROL. The job hangs after the ROL print out to the output file but before the status test. The job will hang at the same point/iteration every time.

For basic debugging I put some print statements at the beginning of the `ROLObjective.update` method. This has suggested that where the job is hanging, some processes have `UpdateType.Revert` flag but others have `UpdateType.Accept`I can’t seem to get the norm to print out with this strategy.

The relevant directory on gadi: `/scratch/xd2/rt8300/02_full_melt_region/trbsht_m0_u1e-1_d1e-2_s1e-1_f0` which has everything to reproduce the problem. An example of the output is in `m0_u1e-1_d1e-2_s1e-1_f0/output_10.dat`. There is also the MPI stack trace in there but from my untrained eyes it doesn't look very helpful.

Just in case this means something - a similarly hanging simulation had a stack trace showing the accept processes with:
```
Thread 1 (Thread 0x145ce88def00 (LWP 671067)):
        #0  0x0000145c9fe82f9b in hmca_bcol_basesmuma_bcast_k_nomial_knownroot_progress () from /apps/hcoll/4.7.3202/lib/hcoll/hmca_bcol_basesmuma.so
        #1  0x0000145caa0e0719 in hmca_coll_ml_parallel_bcast () from /apps/hcoll/4.7.3202/lib/libhcoll.so.1
        #2  0x0000145caa3d1e4f in mca_coll_hcoll_bcast (buff=0x7ffc26131da4, count=1, datatype=0x145cd0685d80 <ompi_mpi_int>, root=0, comm=0x260f590, module=0x39a5e10) at /jobfs/35226956.gadi-pbs/0/openmpi/4.0.7/source/openmpi-4.0.7/ompi/mca/coll/hcoll/coll_hcoll_ops.c:59
```
and the revert with:
```
Thread 1 (Thread 0x149804a45f00 (LWP 671065)):
        #0  0x00001497bbfeb1a0 in hmca_bcol_basesmuma_barrier_fanout_progress_x86 () from /apps/hcoll/4.7.3202/lib/hcoll/hmca_bcol_basesmuma.so
        #1  0x00001497c62f40ec in hmca_coll_ml_barrier_intra () from /apps/hcoll/4.7.3202/lib/libhcoll.so.1
        #2  0x00001497c65efd3a in mca_coll_hcoll_barrier (comm=0x1497ec6f2f40 <ompi_mpi_comm_world>, module=0x21734d0) at /jobfs/35226956.gadi-pbs/0/openmpi/4.0.7/source/openmpi-4.0.7/ompi/mca/coll/hcoll/coll_hcoll_ops.c:29
```
(The full output for this one being in `/scratch/xd2/rt8300/01_melt_region_gadopt/m1e0_u1e0_d0_s0_NoF_trbsht/debug_output_1.dat`)

Hopefully this makes sense. Please let me know what or any info needed from me.

**Comment by angus-g** [comment]
Alright, I think I've figured this one out! I guess it comes down as a bug in the checkpointing in ROL. One of the boolean members of `LinMoreAlgorithm` isn't initialised when restoring from disk, so its value isn't consistent across the processes. On some of them, it ends up calling `TrustRegionUtilities::analyzeRatio` twice: this second call can return a different trust region flag than the first call.

I've pushed a change to the Trilinos fork that pyrol pulls in, which should now restore this parameter. You'll only need to rebuild pyrol in your Firedrake prefix tarball. This will also require some manual intervention on existing checkpoints, but they'll be otherwise compatible. After backing up the checkpoints, the following should work to add the required value so they restore correctly:

```bash
for f in rol_typeb_algorithm_*.cereal; do
  printf '\0\0' >> $f
done
```
